### PR TITLE
Components for drawing polys (-lines and -gons) on a map.

### PR DESCRIPTION
**Pull request by DracoNoir**
See https://developers.google.com/maps/documentation/javascript/reference#Polyline and https://developers.google.com/maps/documentation/javascript/reference#Polygon for more information on these objects.  Note:  This is a refactoring of PR #101 to work with Polymer 1.0.

### CODE REVIEW

**Comment by DracoNoir** [comment]
Rebased to 1.0.4 release.

**Comment by ebidel** [comment]
Think you'll need another rebase. Sorry :\ this is a big CL that I want time to review.

**Comment by DracoNoir** [comment]
Done.  PTAL.

**Comment by DracoNoir** [comment]
Rebased to 1.0.5 release.

**Comment by DracoNoir** [comment]
Rebased to 1.0.6 release.

**Comment by DracoNoir** [comment]
Rebased to latest upstream merge.

**Comment by DracoNoir** [comment]
Rebased to latest upstream merge.

**Comment by DracoNoir** [comment]
ping

**Comment by Lull3rSkat3r** [comment]
+1

**Comment by ebidel** [comment]
Just merged a big PR. This will need another rebase. Sorry!

**Comment by DracoNoir** [comment]
Done.

**Comment by ebidel** [comment]
Taking a look tonight

File: demo/index.html [hunk 1]
```diff
@@ -7,6 +7,7 @@
   <script src="../../webcomponentsjs/webcomponents-lite.min.js"></script>
   <link rel="import" href="../google-map.html">
   <link rel="import" href="../google-map-marker.html">
+  <link rel="import" href="../google-map-poly.html">
```
Comment (ebidel) [inline demo/index.html]: can we make another demo file for this instead of adding it to the basic maps demo?

File: google-map-point.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="../google-apis/google-maps-api.html">
+
+<!--
+The `google-map-point` element represents a point on a map. It's used as a child of other
+google-map-* elements.
+
+<b>Example</b>:
+
+    <google-map-point latitude="37.77493" longitude="-122.41942"></google-map-point>
```
Comment (ebidel) [inline google-map-point.html]: this doesnt show the point you made above, which is that this element needs to be a child of another map element :) Please include.

File: google-map-point.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+-->
+<dom-module id="google-map-point">
+  <style>
+    :host {
+      display: none;
```
Comment (ebidel) [inline google-map-point.html]: no need for `<dom-module>`. Remove it and use `hostAttributes: {hidden: true}` in the prototype.

File: google-map-point.html [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    is: 'google-map-point',
+
+    properties: {
+      /**
+       * The point's longitude coordinate.
+       */
+       longitude: {
+         type: Number,
+         value: null,
+         notify: true,
```
Comment (ebidel) [inline google-map-point.html]: I dont think these need to notify. Nothing inside this element is changing the lat/lng and I wouldnt expect anything to.

File: google-map-point.html [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+         reflectToAttribute: true
```
Comment (ebidel) [inline google-map-point.html]: remove reflectToAttribute on both.

File: google-map-point.html [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+       }
+    },
+
+    /**
+     * Returns the point as a Google Maps LatLng object.
+     *
+     * @method getPosition
```
Comment (ebidel) [inline google-map-point.html]: remove `@method`

File: google-map-point.html [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+     * @return {google.maps.LatLnt} The LatLng object.
+     */
+    getPosition: function() {
+      return new google.maps.LatLng(this.latitude, this.longitude);
```
Comment (ebidel) [inline google-map-point.html]: I've been told a literal is better now. 

`return {lat: this.latitude, lng: this.longitude}`

File: google-map-poly.html [hunk 8]
```diff
@@ -0,0 +1,554 @@
+<!-- Copyright (c) 2015 Google Inc. All rights reserved. -->
+
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="../google-apis/google-maps-api.html">
+<link rel="import" href="google-map-point.html">
+
+<!--
+The `google-map-poly` element represents a series of connected line segments (aka a polyline) which
```
Comment (ebidel) [inline google-map-poly.html]: `google-map-polyline` might make more sense.

File: google-map-poly.html [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+      <google-map-poly>
+        <google-map-point latitude="37.77493" longitude="-122.41942"></google-map-point>
+        <google-map-point latitude="38.77493" longitude="-123.41942"></google-map-point>
+      </google-map-poly>
+    </google-map>
+
+<b>Example</b>—a semi-translucent blue triangle:
+
+    <google-map latitude="37.77493" longitude="-122.41942">
+      <google-map-poly closed fillColor="blue" fillOpacity=".5">
```
Comment (ebidel) [inline google-map-poly.html]: `fill-color`, `fill-opacity`

File: google-map-poly.html [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+        <google-map-point latitude="36.77493" longitude="-121.41942"></google-map-point>
+        <google-map-point latitude="38.77493" longitude="-122.41942"></google-map-point>
```
Comment (ebidel) [inline google-map-poly.html]: are points used other places in the maps api? i.e. do we need a DOM element for google-map-point, or should the polyline element have an array property, `points`? 

cc @brendankenny

File: google-map-poly.html [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+        <google-map-point latitude="36.77493" longitude="-123.41942"></google-map-point>
+      </google-map-poly>
+    </google-map>
+
+@element google-map-poly
+@status alpha
```
Comment (ebidel) [inline google-map-poly.html]: remove all of these old doc pragmas

File: google-map-poly.html [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+        observer: '_closedChanged'
+      },
+
+      /**
+       * When true, the poly may be dragged to a new position.
+       */
+      draggable: {
+        type: Boolean,
+        value: false,
+        //observer: '_draggableChanged'
```
Comment (ebidel) [inline google-map-poly.html]: remove

File: google-map-poly.html [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+       * If the path is not closed, the icons to be rendered along the polyline.
+       */
+      icons: {
+        type: Array,
+        value: null,
+        observer: '_iconsChanged'
+      },
+
+      /**
+       * When true, poly mouse* events will be automatically registered.
```
Comment (ebidel) [inline google-map-poly.html]: mouse*?

File: google-map-poly.html [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+        type: Number,
+        value: 0,
+        observer: '_zIndexChanged'
+      }
+    },
+
+    // Lifecycle event handlers.
+
+    detached: function() {
+      this.poly.setMap(null);
```
Comment (ebidel) [inline google-map-poly.html]: how about event listener cleanup?

File: google-map-poly.html [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+
+    attached: function() {
+      // If element is added back to DOM, put it back on the map.
+      this.poly && this.poly.setMap(this.map);
+    },
+
+    // Attribute/property change watchers.
+
+    attributeChanged: function(attrName, oldVal, newVal) {
+      if (!this.poly) return;
```
Comment (ebidel) [inline google-map-poly.html]: brackets please

File: google-map-poly.html [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+
+    _zIndexChanged: function() {
+      this.poly && this.poly.set('zIndex', this.zIndex);
+    },
+
+    // Helper logic.
+
+    _buildPathFromPoints: function() {
+      this._points = Array.prototype.slice.call(Polymer.dom(this.$.points).getDistributedNodes());
+      window.console.log("Points:  " + this._points.length);
```
Comment (ebidel) [inline google-map-poly.html]: remove the logs

File: google-map-poly.html [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+      window.console.log(Polymer.dom(this).getDestinationInsertionPoints());
+
+      // Build path from current points (ignoring vertex insertions while doing so).
+      this._building = true;
```
Comment (ebidel) [inline google-map-poly.html]: Do you really need this? This operation should be super fast and you're only using this before when someone tries to edit.

File: google-map-poly.html [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+      this._building = false;
+
+      this.fire('google-map-poly-path-built', this.path);
+
+      // Watch for future updates.
+      if (this._mutationObserver) {
+        return;
+      }
+      this._mutationObserver = new MutationObserver(this._buildPathFromPoints.bind(this));
+      this._mutationObserver.observe(this, {
```
Comment (ebidel) [inline google-map-poly.html]: `Polymer.dom(this)`

File: google-map-poly.html [hunk 19]
```diff
@@ -0,0 +1,554 @@
...
+
+    _clearListener: function(name) {
+      if (this._listeners[name]) {
+        google.maps.event.removeListener(this._listeners[name]);
+        this._listeners[name] = null;
+      }
+    },
+
+    _convertStrokePosition: function() {
+      return google.maps.StrokePosition && this.strokePosition?
```
Comment (ebidel) [inline google-map-poly.html]: space: `this.strokePosition ?`

File: google-map-poly.html [hunk 20]
```diff
@@ -0,0 +1,554 @@
...
+          google.maps.StrokePosition[this.strokePosition.toUpperCase()] : 0;
+    },
+
+    _createPoly: function() {
+      // Build poly's path and register mutation listeners on first creation.
```
Comment (ebidel) [inline google-map-poly.html]: s/listeners/observer

File: google-map-poly.html [hunk 21]
```diff
@@ -0,0 +1,554 @@
...
+      if (!this.path) {
+        this._setPath(new google.maps.MVCArray());
+        google.maps.event.addListener(this.path, 'insert_at', this._startEditing.bind(this));
+        google.maps.event.addListener(this.path, 'set_at', this._updatePoint.bind(this));
+        this._buildPathFromPoints();
```
Comment (ebidel) [inline google-map-poly.html]: seems like you could create the MO here and not have to deal with it in buildPathFromPoints. your comment above also describes it being here.

File: google-map-poly.html [hunk 22]
```diff
@@ -0,0 +1,554 @@
...
+        options.fillColor = this.fillColor;
+        options.fillOpacity = this.fillOpacity;
+        this._setPoly(new google.maps.Polygon(options));
+      } else {
+        options.icons = this.icons;
+        this._setPoly(new google.maps.Polyline(options));
+      }
+
+      this._listeners = {};
+      this._clickEventsChanged();
```
Comment (ebidel) [inline google-map-poly.html]: Check that these don't double call on page load. e.g. someone uses the `click-events` attribute. Won't this call and the natural `_clickEventsChanged()` setup double listeners?

File: google-map-poly.html [hunk 23]
```diff
@@ -0,0 +1,554 @@
...
+    },
+
+    _forwardEvent: function(name) {
+      this._listeners[name] = google.maps.event.addListener(this.poly, name, function(event) {
+        this.fire('google-map-poly-' + name, event);
+      }.bind(this));
+    },
+
+    _startEditing: function(index) {
+      if (this._building) return;  // Ignore changes while building path.
```
Comment (ebidel) [inline google-map-poly.html]: brackets please

File: google-map-poly.html [hunk 24]
```diff
@@ -0,0 +1,554 @@
...
+        google.maps.event.addListenerOnce(this.map, 'click', function() {
+          this._setEditing(false);
+          this.fire('google-map-poly-path-updated', this.path);
+        }.bind(this));
+      }
+    },
+
+    _updatePoint: function(index, vertex) {
+      // Ignore changes if path is out of sync with google-map-point elements.
+      if (!this._points) return;
```
Comment (ebidel) [inline google-map-poly.html]: brackets

File: google-map.html [hunk 25]
```diff
@@ -365,6 +366,10 @@
       '_debounceUpdateCenter(latitude, longitude)'
     ],
 
+    created: function() {
+      this.objects = [];
```
Comment (ebidel) [inline google-map.html]: in 1.0, this should be done in the prototype: `value: function() return []; }`

File: google-map.html [hunk 26]
```diff
@@ -469,6 +479,45 @@
...
+          return;
+        }
+      }
+
+      this._observeObjects();
+
+      this.objects = newObjects;
+
+      // Set the map property on each object.
+      if (this.objects.length && this.map) {
```
Comment (ebidel) [inline google-map.html]: we could probably make this a helper. `_observeMarkers` does the same.

**Comment by ebidel** [comment]
Reviewed

File: google-map-poly.html (hunk 20 above)
Comment (DracoNoir) [inline google-map-poly.html]: Listener is the terminology that the Maps API uses for the function that's called when a map object event is fired.

File: google-map-poly.html (hunk 21 above)
Comment (DracoNoir) [inline google-map-poly.html]: The MO is tied to the array of google-map-point components (not the Maps MVCArray), so it seems better to keep the logic together.

File: google-map-poly.html (hunk 18 above)
Comment (DracoNoir) [inline google-map-poly.html]: That doesn't work:
  Uncaught NotFoundError: Failed to execute 'observe' on 'MutationObserver': The provided node was null.

File: google-map-poly.html (hunk 17 above)
Comment (DracoNoir) [inline google-map-poly.html]: Yes.  This prevents the component from mistaking the building of the path from the DOM as the beginning of user edits (both fire the 'insert_at' event to the path MVCArray's listener).

File: google-map-poly.html (hunk 10 above)
Comment (DracoNoir) [inline google-map-poly.html]: Yes.  All map objects require at least one point and a few (polygons, polylines, and rectangles) require multiple.  I thought it was natural to use an element, but I could convert it to a property if that is more Polymer-y.

File: google-map-poly.html (hunk 8 above)
Comment (DracoNoir) [inline google-map-poly.html]: This supports both polygons and polylines.  Calling it just polyline might confuse people familiar with the raw API and creating separate components would effectively duplicate most of the code so I thought it efficient to shorten the name and have a single implementation with a flag.

File: google-map-point.html (hunk 7 above)
Comment (DracoNoir) [inline google-map-point.html]: This doesn't actually work everywhere:
  Assertion failed: InvalidValueError: at index 0: not an instance of LatLng

File: google-map-point.html (hunk 2 above)
Comment (DracoNoir) [inline google-map-point.html]: done

File: google-map-point.html (hunk 3 above)
Comment (DracoNoir) [inline google-map-point.html]: done

File: google-map-point.html (hunk 4 above)
Comment (DracoNoir) [inline google-map-point.html]: done

File: google-map-point.html (hunk 5 above)
Comment (DracoNoir) [inline google-map-point.html]: done

File: google-map-point.html (hunk 6 above)
Comment (DracoNoir) [inline google-map-point.html]: done

File: google-map-poly.html (hunk 9 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 11 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 12 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 13 above)
Comment (DracoNoir) [inline google-map-poly.html]: expanded the description

File: google-map-poly.html (hunk 14 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

Comment (DracoNoir) [inline google-map-poly.html]: Note:  google-map-marker has the same problem.

File: google-map-poly.html (hunk 15 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 16 above)
Comment (DracoNoir) [inline google-map-poly.html]: Oops.  Done.

File: google-map-poly.html (hunk 19 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 22 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

Comment (DracoNoir) [inline google-map-poly.html]: Note:  google-map-marker has the same problem.

File: google-map-poly.html (hunk 23 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map-poly.html (hunk 24 above)
Comment (DracoNoir) [inline google-map-poly.html]: done

File: google-map.html (hunk 25 above)
Comment (DracoNoir) [inline google-map.html]: done

File: google-map.html (hunk 26 above)
Comment (DracoNoir) [inline google-map.html]: done

**Comment by DracoNoir** [comment]
PTAL

File: google-map-poly.html (hunk 8 above)
Comment (ebidel) [inline google-map-poly.html]: Ack

File: google-map-poly.html (hunk 14 above)
Comment (ebidel) [inline google-map-poly.html]: Mind filing a bug?

**Comment by ebidel** [comment]
LGTM.  A couple of more nits.

**Comment by DracoNoir** [comment]
PTAL

**Comment by ebidel** [comment]
LGTM!

**Comment by Lull3rSkat3r** [comment]
:clap:

**Comment by ebidel** [comment]
We should probably support a points property on the mask element for cases
where users have 1000s of points. Theres a similar markers property for
that reason.

**Pull request merged**